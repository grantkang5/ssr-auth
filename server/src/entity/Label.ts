// import {
//   BaseEntity,
//   Entity,
//   PrimaryGeneratedColumn,
//   Column,
//   CreateDateColumn,
//   UpdateDateColumn,
//   ManyToOne,
//   ManyToMany,
//   JoinTable
// } from 'typeorm';
// import { ObjectType, Field, Int } from 'type-graphql';
// import { Project } from './Project';
// import { Task } from './Task';

// @ObjectType()
// @Entity('labels')
// export class Label extends BaseEntity {
//   @Field(() => Int)
//   @PrimaryGeneratedColumn()
//   id: number;

//   @Field()
//   @Column()
//   name: string;

//   @Field()
//   @Column()
//   color: string;

//   @CreateDateColumn()
//   created_at: Date;

//   @UpdateDateColumn()
//   updated_at: Date;

//   @ManyToOne(() => Project, project => project.labels)
//   project: Project;

//   @ManyToMany(() => Task)
//   @JoinTable()
//   tasks: Task[];
// }
