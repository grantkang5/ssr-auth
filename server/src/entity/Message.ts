// import {
//   BaseEntity,
//   Entity,
//   PrimaryGeneratedColumn,
//   Column,
//   CreateDateColumn,
//   UpdateDateColumn,
//   ManyToOne
// } from 'typeorm';
// import { ObjectType, Field, Int } from 'type-graphql';
// import { User } from './User';
// import { Task } from './Task';

// @ObjectType()
// @Entity('messages')
// export class Message extends BaseEntity {
//   @Field(() => Int)
//   @PrimaryGeneratedColumn()
//   id: number;

//   @Field()
//   @Column()
//   message: string;

//   @CreateDateColumn()
//   created_at: Date;

//   @UpdateDateColumn()
//   updated_at: Date;

//   @ManyToOne(() => User, user => user.messages)
//   user: User;

//   @ManyToOne(() => Task, task => task.messages)
//   task: Task;
// }
