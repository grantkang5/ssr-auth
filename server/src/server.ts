import { ApolloServer } from "apollo-server-express";
import { buildSchemaSync } from "type-graphql";
import { UserResolver } from "./resolvers/UserResolver";

export const server = new ApolloServer({
  schema: buildSchemaSync({ resolvers: [UserResolver] }),
  context: ({ req, res }) => ({ req, res })
});