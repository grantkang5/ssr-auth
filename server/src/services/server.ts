import { ApolloServer } from 'apollo-server-express';
import { buildSchemaSync } from 'type-graphql';
import { pubSub } from './redis';

import resolvers from '../resolvers';

export const server = new ApolloServer({
  schema: buildSchemaSync({
    resolvers,
    pubSub
  }),
  context: ({ req, res }) => ({ req, res })
});
