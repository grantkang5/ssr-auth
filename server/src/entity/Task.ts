import { ObjectType, Field, ID } from "type-graphql";
import { Entity, BaseEntity, PrimaryGeneratedColumn, Column, CreateDateColumn, UpdateDateColumn, ManyToOne, BeforeInsert } from "typeorm";
import { List } from "./List";

@ObjectType()
@Entity('tasks')
export class Task extends BaseEntity {
  @BeforeInsert()
  async generatePos() {
    const maxPosTask = await Task.findOne({
      order: {
        id: "DESC"
      },
      where: {
        list: this.list.id
      }
    })
    this.pos = maxPosTask ? maxPosTask.pos + 16384 : 16384
  }

  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: number;

  @Field()
  @Column()
  name: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  desc: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  dueDate: Date;

  @Field()
  @Column({ type: 'double precision' })
  pos: number;

  @CreateDateColumn()
  created_at: Date;

  @UpdateDateColumn()
  updated_at: Date;

  @Field(() => List)
  @ManyToOne(() => List, list => list.tasks, {
    cascade: ["insert", "update"],
    onDelete: 'CASCADE',
    nullable: false
  })
  list: List
}
