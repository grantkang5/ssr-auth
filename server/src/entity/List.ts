import {
  BaseEntity,
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  BeforeInsert,
  OneToMany
} from 'typeorm';
import { ObjectType, Field, ID } from 'type-graphql';
import { Project } from './Project';
import { Task } from './Task';

@ObjectType()
@Entity('lists')
export class List extends BaseEntity {
  @BeforeInsert()
  async generatePos() {
    try {
      this.project.maxPos += 16384;
      await this.project.save();
      this.pos = this.project.maxPos;
    } catch (err) {
      console.log(err);
      return err;
    }
  }
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: number;

  @Field()
  @Column()
  name: string;

  @Field()
  @Column({ type: 'double precision' })
  pos: number;

  @CreateDateColumn()
  created_at: Date;

  @UpdateDateColumn()
  updated_at: Date;

  @Field(() => Project)
  @ManyToOne(() => Project, project => project.lists, {
    onDelete: 'CASCADE',
    nullable: false
  })
  project: Project;

  @Field(() => [Task])
  @OneToMany(() => Task, task => task.list)
  tasks: Task[];
}
