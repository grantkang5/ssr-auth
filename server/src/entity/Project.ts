import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  BaseEntity,
  ManyToMany,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  OneToMany
} from 'typeorm';
import { ObjectType, Field, Int } from 'type-graphql';
import { User } from './User';
import { List } from './List';
import { Team } from './Team';

@ObjectType()
@Entity('projects')
export class Project extends BaseEntity {
  @Field(() => Int)
  @PrimaryGeneratedColumn()
  id: number;

  @Field()
  @Column()
  name: string;

  @Field({ nullable: true })
  @Column({ nullable: true })
  desc: string;

  @Column({ type: 'double precision', default: 0 })
  maxPos: number;

  @Field()
  @CreateDateColumn()
  created_at: Date;

  @Field()
  @UpdateDateColumn()
  updated_at: Date;

  @Field(() => User)
  @ManyToOne(() => User, user => user.ownedProjects, {
    onDelete: 'CASCADE'
  })
  owner: User;

  @Field(() => [List])
  @OneToMany(() => List, list => list.project, {
    cascade: true,
    eager: true
  })
  lists: List[];

  @Field(() => [User])
  @ManyToMany(() => User, user => user.projects)
  members: User[];

  @Field(() => Team)
  @ManyToOne(() => Team, team => team.projects, {
    onDelete: "SET NULL"
  })
  team: Team;
}
