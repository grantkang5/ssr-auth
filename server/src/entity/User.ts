import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  BaseEntity,
  // ManyToMany,
  // JoinTable,
  OneToMany,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToMany,
  JoinTable
} from 'typeorm';
import { ObjectType, Field, ID } from 'type-graphql';
import { Project } from './Project';
import { Team } from './Team';

@ObjectType()
@Entity('users')
export class User extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: number;

  @Field()
  @Column({ unique: true })
  email: string;

  @Column({ nullable: true })
  password: string;

  @Field()
  @Column()
  username: string;

  @Field()
  @Column({ nullable: true })
  avatar: string;

  @Column('int', { default: 0 })
  tokenVersion: number;
  // TODO: make enum. 'website' | 'google'

  @Field()
  @Column({ default: 'website' })
  auth: string;

  @Field()
  @CreateDateColumn()
  created_at: Date;

  @Field()
  @UpdateDateColumn()
  updated_at: Date;

  @Field(() => [Project])
  @OneToMany(() => Project, project => project.owner, {
    cascade: true,
    eager: true
  })
  ownedProjects: Project[];

  @Field(() => [Project])
  @ManyToMany(() => Project, project => project.members)
  @JoinTable()
  projects: Project[];

  @Field(() => [Team])
  @ManyToMany(() => Team, team => team.members)
  @JoinTable()
  teams: Team[];
}
