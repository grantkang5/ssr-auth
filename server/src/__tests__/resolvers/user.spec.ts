import { gql } from "apollo-server-express";
import { testServer, initDb } from "../mocks/server";
import { createConnection, Connection } from "typeorm";
import { User } from "../../entity/User";

const { createTestClient } = require("apollo-server-testing");
const { query, mutate } = createTestClient(testServer);

describe("User Resolver", () => {
  let connection: Connection;
  beforeAll(async () => {
    initDb();
    connection = await createConnection();
  });

  afterAll(() => {
    connection.close();
  });

  describe("Me query", () => {
    it("should fetch a current user", async () => {
      const res = await query({
        query: gql`
          query Me {
            me {
              id
              email
            }
          }
        `
      });
      expect(res.data).toBeDefined();
      expect(res.errors).toBeUndefined();
      expect(res.data).toMatchSnapshot();
    });
  });

  describe("Login mutation", () => {
    it("should return an accessToken", async () => {
      const res = await mutate({
        mutation: gql`
          mutation Login($email: String!, $password: String!) {
            login(email: $email, password: $password) {
              accessToken
            }
          }
        `,
        variables: { email: "dev@email.com", password: "password" }
      });
      expect(res.data).toBeDefined();
      expect(res.errors).toBeUndefined();
    });
  });

  describe("Register mutation", () => {
    it("should register a user into the db", async () => {
      const res = await mutate({
        mutation: gql`
          mutation Register($email: String!, $password: String!) {
            register(email: $email, password: $password) {
              id
              email
            }
          }
        `,
        variables: { email: "dev2@email.com", password: "password" }
      });
      const user = await User.findOne({ email: "dev2@email.com" })
      expect(user!.email).toEqual('dev2@email.com')
      expect(res.data).toBeDefined()
      expect(res.errors).toBeUndefined()
      expect(res.data).toMatchSnapshot();
    });
  });

  describe("Logout mutation", () => {
    it("should log the user out", async () => {
      const res = await mutate({
        mutation: gql`
          mutation Logout {
            logout
          }
        `
      })
      expect(res.data).toBeDefined()
      expect(res.errors).toBeUndefined()
      expect(res.data).toMatchSnapshot();
    })
  })
});
