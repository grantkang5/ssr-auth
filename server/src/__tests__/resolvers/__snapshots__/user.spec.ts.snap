// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Resolver Login mutation should return an accessToken 1`] = `
Object {
  "login": Object {
    "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjEsImlhdCI6MTU3MDg0ODY3NSwiZXhwIjoxNTcwODQ5NTc1fQ.zK0arbMJoRUFpp2c0mb8QxTvuaK44z_bh6Q91r4cZw8",
  },
}
`;

exports[`User Resolver Logout mutation should log the user out 1`] = `
Object {
  "logout": true,
}
`;

exports[`User Resolver Me query should fetch a current user 1`] = `
Object {
  "me": Object {
    "email": "dev@email.com",
    "id": 1,
  },
}
`;

exports[`User Resolver Mutations should login a user 1`] = `
Object {
  "login": Object {
    "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOjEsImlhdCI6MTU3MDg0ODU3MCwiZXhwIjoxNTcwODQ5NDcwfQ.vdJ5N1jZj2y4vpFEHOEoRbuVC7LrJb5TZVsxjB-Lkwk",
  },
}
`;

exports[`User Resolver Queries should fetch a current user 1`] = `
Object {
  "me": Object {
    "email": "dev",
    "id": 1,
  },
}
`;

exports[`User Resolver Register mutation should register a user into the db 1`] = `
Object {
  "register": Object {
    "email": "dev2@email.com",
    "id": 11,
  },
}
`;
